lib/ansible/modules/cloud/amazon/ec2_asg.py

```diff
@@ -441,173 +510,272 @@ def create_autoscaling_group(connection, module):
+                    ]
+                )
+            as_group = connection.describe_auto_scaling_groups(
+                AutoScalingGroupNames=[group_name])['AutoScalingGroups'][0]
             asg_properties = get_properties(as_group)
             changed = True
-            return(changed, asg_properties)
-        except BotoServerError as e:
+            return changed, asg_properties
+        except botocore.exceptions.BotoCoreError as e:
             module.fail_json(msg="Failed to create Autoscaling Group: %s" % str(e), exception=traceback.format_exc(e))
     else:
-        as_group = as_groups[0]
+        as_group = as_groups['AutoScalingGroups'][0]
         changed = False
 
-        if suspend_processes(as_group, module):
+        if suspend_processes(connection, as_group, module):
             changed = True
 
-        for attr in ASG_ATTRIBUTES:
-            if module.params.get(attr, None) is not None:
-                module_attr = module.params.get(attr)
-                if attr == 'vpc_zone_identifier':
-                    module_attr = ','.join(module_attr)
-                group_attr = getattr(as_group, attr)
-                # we do this because AWS and the module may return the same list
-                # sorted differently
-                if attr != 'termination_policies':
-                    try:
-                        module_attr.sort()
-                    except:
-                        pass
-                    try:
-                        group_attr.sort()
-                    except:
-                        pass
-                if group_attr != module_attr:
-                    changed = True
-                    setattr(as_group, attr, module_attr)
-
+        # process tag changes
         if len(set_tags) > 0:
-            have_tags = {}
-            want_tags = {}
-
-            for tag in asg_tags:
-                want_tags[tag.key] = [tag.value, tag.propagate_at_launch]
-
+            have_tags = as_group.get('Tags')
+            want_tags = asg_tags
             dead_tags = []
-            for tag in as_group.tags:
-                have_tags[tag.key] = [tag.value, tag.propagate_at_launch]
-                if tag.key not in want_tags:
+            have_tag_keyvals = [x['Key'] for x in have_tags]
+            want_tag_keyvals = [x['Key'] for x in want_tags]
+            for have_tag_keyval in have_tag_keyvals:
+                if have_tag_keyval not in want_tag_keyvals:
                     changed = True
-                    dead_tags.append(tag)
-
-            if dead_tags != []:
-                connection.delete_tags(dead_tags)
-
-            if have_tags != want_tags:
+                    dead_tags.append(dict(ResourceId=as_group['AutoScalingGroupName'],
+                                          ResourceType='auto-scaling-group', Key=have_tag_keyval))
+                    have_tags = [have_tag for have_tag in have_tags if have_tag['Key'] != have_tag_keyval]
```

Comment:
This is a bit confusing since it's assigning to the same name name as comes in. I think this section could be cleared up with the use of `set`. So something along the lines of:

```
for dead_tag in set(have_tag_keyvals).difference(want_tag_keyvals):
    dead_tags.append(..... and so on ...)
```